File: Solve-Path/Solve-Path/src/main/java/com/example/solvePath/Node.java
package com.example.solvePath;

import java.util.Arrays;
import java.util.Vector;

public class Node {
    public State state;
    public int f; // Tổng chi phí
    public int g; // Chi phí từ node gốc đến node này
    public int h; // Chi phí ước lượng từ node này đến node mục tiêu
    public int cost; // Chi phí đi vào node này
    public Node parent; // Node cha

    // Phương thức khởi tạo
    public Node(State state, int cost) {
        this.state = state;
        this.cost = cost;
    }

    // Kiểm tra value có giống value của n không
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Node n = (Node) obj;
        return Arrays.equals(state.value, n.state.value);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(state.value);
    }

    // Tính ước lượng h(x)
    public int estimate(State goalState) {
        return state.estimate(goalState);
    }

    // Vector các Node con
    public Vector<Node> successors() {
        Vector<Node> nodes = new Vector<>();
        Vector<State> states = state.successors();
        for (State value : states) {
            nodes.add(new Node(value, cost + 1)); // Tăng chi phí cho mỗi bước
        }
        return nodes;
    }

    // Tính toán giá trị f
    public void calculateF() {
        f = g + h;
    }
}

File: NPuzzleAI/NPuzzleAI/Node.cs
using System.Collections.Generic;

namespace NPuzzleAI
{
    public class Node
    {
        public State State { get; set; }
        public int F { get; set; }
        public int G { get; set; }
        public int H { get; set; }
        public int Cost { get; set; }
        public Node Parent { get; set; }

        // Phương thức khởi tạo
        public Node(State state, int cost)
        {
           this.State = state;
            this.Cost = cost;
        }



        // Kiểm tra value có giống value của n không
        public bool Equals(Node n)
        {
            bool flag = true;
            int[] val = State.Value;
            int[] newVal = n.State.Value;
            for (int i = 0; i < val.Length; i++)
            {
                if (val[i] != newVal[i])
                {
                    flag = false;
                    break;
                }
            }
            return flag;
        }

        // Tính ước lượng h(x)
        public int Estimate(State goalState)
        {
            return State.Estimate(goalState);
        }

        // Vector các Node con
        public List<Node> Successors()
        {
            List<Node> nodes = new List<Node>();
            List<State> states = State.Successors();
            foreach (State value in states)
            {
                nodes.Add(new Node(value, 1));
            }
            return nodes;
        }
    }
}
